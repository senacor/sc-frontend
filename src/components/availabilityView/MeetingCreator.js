import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles/index';
import * as actions from '../../actions';
import { connect } from 'react-redux';
import { getSelectedDate, getMeeting } from '../../reducers/selector';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import moment from 'moment-timezone';
import AppointmentPicker from './AppointmentPicker';

const styles = theme => ({
  container: {
    '& > *': {
      // display: 'block !important',
      marginLeft: theme.spacing.unit,
      marginRight: theme.spacing.unit
    }
  },
  attendeeList: {
    '& ListItemText span': {
      fontSize: '0.875rem'
    }
  }
});

class MeetingView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      location: '',
      date: null,
      startTime: null,
      endTime: null
    };
  }

  handleChange = name => event => {
    this.setState({
      [name]: event.target.value
    });
  };

  handleClickOfMeetingButton = event => {
    event.preventDefault();
    this.addMeeting();
  };

  setStartAndEndTime = (date, startTime, endTime) => {
    this.setState({
      date: date,
      startTime: startTime,
      endTime: endTime
    });
  };

  addMeeting() {
    let startDateTime = moment(`${this.state.date} ${this.state.startTime}`).tz(
      'Europe/Berlin'
    );
    let endDateTime = moment(`${this.state.date} ${this.state.endTime}`).tz(
      'Europe/Berlin'
    );
    let meeting_details = {
      //TODO replace hardcoded values
      prId: 1,
      start: startDateTime.utc().format('YYYY-MM-DDTHH:mmZ'),
      end: endDateTime.utc().format('YYYY-MM-DDTHH:mmZ'),
      location: this.state.location,
      //TODO replace hardcoded values
      requiredAttendeeIds: [4],
      optionalAttendeeIds: []
    };
    this.props.addMeeting(meeting_details);
  }

  render() {
    const { classes } = this.props;
    return (
      <div>
        <AppointmentPicker onTimeChange={this.setStartAndEndTime} />
        <form className={classes.container} noValidate autoComplete="off">
          <TextField
            id="location"
            label="Ort"
            className={classes.textField}
            value={this.state.location}
            onChange={this.handleChange('location')}
            margin="normal"
          />
          <Button
            id="createMeeting"
            className={classes.button}
            variant="raised"
            onClick={this.handleClickOfMeetingButton}
            type="submit"
          >
            Termin erstellen
          </Button>
        </form>
      </div>
    );
  }
}

MeetingView.propTypes = {
  classes: PropTypes.object.isRequired
};

export const StyledComponent = withStyles(styles)(MeetingView);
export default connect(
  state => ({
    meeting: getMeeting(state),
    getSelectedDateTime: getSelectedDate(state)
  }),
  {
    addMeeting: actions.addMeeting
  }
)(StyledComponent);
