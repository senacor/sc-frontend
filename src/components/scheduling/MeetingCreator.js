import React, { useState, useEffect, useContext } from 'react';
import ObjectGet from 'object-get';
import { withStyles } from '@material-ui/core';
import Typography from '@material-ui/core/Typography/Typography';
import Grid from '@material-ui/core/Grid/Grid';

import TimeTable from './AppointmentTable/TimeTable';
import Attendee from './AppointmentTable/Attendee';
import MeetingCreatorForm from './MeetingCreatorForm';
import PersonToggle from './PersonToggle';
import { extractAppointments } from './AppointmentTable/AppointmentUtilities';
import { hasRoleInPrBasedOnUserName } from '../../helper/checkRole';
import meetingDetailVisibilityService from '../../service/MeetingDetailVisibilityService';
import PrStatusActionButton from '../pr/PrStatusActionButton';
import { ErrorContext, MeetingContext, UserinfoContext } from '../App';
import { appointmentsSearch } from '../../calls/meetings';

const styles = theme => ({
  title: {
    fontSize: 20,
    paddingBottom: 3 * theme.spacing.unit
  }
});

export const MeetingCreator = ({
  classes,
  intl,
  pr,
  selectedDate,
  handleChange
}) => {
  const { userroles, userinfo } = useContext(UserinfoContext.context).value;
  const { value: meeting } = useContext(MeetingContext.context);
  const [employee, setEmployee] = useState('');
  const [supervisor, setSupervisor] = useState('');
  const [reviewer, setReviewer] = useState('');
  const [appointmentResults, setAppointmentResults] = useState([]);

  let errorContext = useContext(ErrorContext.context);

  const setStateforRole = (pr, role, userinfo) => {
    let hasRoleInPr = hasRoleInPrBasedOnUserName(pr, userinfo);
    let roleDescription = '';
    if (hasRoleInPr([role])) {
      roleDescription = `${intl.formatMessage({
        id: 'meetingcreator.me'
      })}`;
    } else {
      switch (role) {
        case 'employee':
          roleDescription = `${intl.formatMessage({
            id: 'meetingcreator.employee'
          })}`;
          break;
        case 'supervisor':
          roleDescription = `${intl.formatMessage({
            id: 'meetingcreator.supervisor'
          })}`;
          break;
        case 'reviewer':
          roleDescription = `${intl.formatMessage({
            id: 'meetingcreator.reviewer'
          })}`;
          break;
        default:
          roleDescription = '';
      }
    }
    switch (role) {
      case 'employee':
        setEmployee(
          Object.assign({}, employee, {
            id: ObjectGet(pr, `${role}.login`),
            name:
              ObjectGet(pr, `${role}.firstName`) +
              ' ' +
              ObjectGet(pr, `${role}.lastName`),
            role: roleDescription,
            show: true
          })
        );
        break;
      case 'supervisor':
        setSupervisor(
          Object.assign({}, supervisor, {
            id: ObjectGet(pr, `${role}.login`),
            name:
              ObjectGet(pr, `${role}.firstName`) +
              ' ' +
              ObjectGet(pr, `${role}.lastName`),
            role: roleDescription,
            show: true
          })
        );
        break;
      case 'reviewer':
        setReviewer(
          Object.assign({}, reviewer, {
            id: ObjectGet(pr, `${role}.login`),
            name:
              ObjectGet(pr, `${role}.firstName`) +
              ' ' +
              ObjectGet(pr, `${role}.lastName`),
            role: roleDescription,
            show: true
          })
        );
        break;
      default:
    }
  };

  const hasSupervisorEntry = pr => {
    return pr.supervisor !== undefined && pr.supervisor.id !== '';
  };

  const hasReviewerEntryThatIsDifferentFromSupervisor = pr => {
    return (
      pr.reviewer !== undefined &&
      pr.reviewer.id !== '' &&
      pr.reviewer.id !== pr.supervisor.id
    );
  };

  const setEmployeeSupervisorReviewerData = (pr, userinfo) => {
    setStateforRole(pr, 'employee', userinfo);
    hasSupervisorEntry(pr) && setStateforRole(pr, 'supervisor', userinfo);
    hasReviewerEntryThatIsDifferentFromSupervisor(pr) &&
      setStateforRole(pr, 'reviewer', userinfo);
  };

  const fetchAppointments = date => {
    let attendees = [pr.employee.login, pr.supervisor.login];
    hasReviewerEntryThatIsDifferentFromSupervisor(pr) &&
      attendees.push(pr.reviewer.login);

    appointmentsSearch(
      attendees.join(','),
      date,
      errorContext,
      setAppointmentResults
    );
  };

  useEffect(() => {
    setEmployeeSupervisorReviewerData(pr, userinfo);
    fetchAppointments(selectedDate);
  }, []);

  const onVisibilityChange = attendee => () => {
    switch (attendee) {
      case 'employee':
        setEmployee(
          Object.assign({}, employee, {
            show: !employee.show
          })
        );
        break;
      case 'supervisor':
        setSupervisor(
          Object.assign({}, supervisor, {
            show: !supervisor.show
          })
        );
        break;
      case 'reviewer':
        setReviewer(
          Object.assign({}, reviewer, {
            show: !reviewer.show
          })
        );
        break;
      default:
    }
  };

  const getStateOfAttendee = attendee => {
    if (attendee === 'employee') {
      return employee;
    } else if (attendee === 'supervisor') {
      return supervisor;
    } else {
      return reviewer;
    }
  };

  const filterConditionWithoutAppointments = key => {
    if (key === 'employee') {
      return employee && employee.id;
    } else if (key === 'supervisor') {
      return supervisor && supervisor.id;
    } else {
      return reviewer && reviewer.id;
    }
  };

  const filterConditionWithAppointments = key => {
    if (key === 'employee') {
      return employee && employee.id && appointmentResults[employee.id];
    } else if (key === 'supervisor') {
      return supervisor && supervisor.id && appointmentResults[supervisor.id];
    } else {
      return reviewer && reviewer.id && appointmentResults[reviewer.id];
    }
  };

  let visibilityService = new meetingDetailVisibilityService();
  visibilityService.setPr(pr);
  visibilityService.setUserinfo(userinfo);
  visibilityService.setUserroles(userroles);
  visibilityService.setMeeting(meeting);
  return (
    <React.Fragment>
      <Typography variant="body1" className={classes.title}>
        {intl.formatMessage({
          id: 'meetingcreator.datescheduling'
        })}
      </Typography>
      <Grid id={'tableRolePick'} container spacing={24} direction="column">
        <Grid item>
          <MeetingCreatorForm
            prById={pr}
            fetchAppointments={fetchAppointments}
          />
        </Grid>
        {visibilityService.getAction() ? (
          <Grid item>
            <Grid container direction="column">
              {['employee', 'supervisor', 'reviewer']
                .filter(key => {
                  return filterConditionWithoutAppointments(key);
                })
                .map(attendee => {
                  return (
                    <Grid item key={attendee}>
                      <PersonToggle
                        displayName={`${getStateOfAttendee(attendee).name}`}
                        displayRole={`${getStateOfAttendee(attendee).role}`}
                        onChange={onVisibilityChange(attendee)}
                        showAttendee={getStateOfAttendee(attendee).show}
                        attendee={attendee}
                      />
                    </Grid>
                  );
                })}
            </Grid>
          </Grid>
        ) : null}
        {visibilityService.getAction() ? (
          <Grid item>
            <TimeTable>
              {['employee', 'supervisor', 'reviewer']
                .filter(key => filterConditionWithAppointments(key))
                .map((attendee, index, keyArray) => (
                  <Attendee
                    key={attendee}
                    show={getStateOfAttendee(attendee).show}
                    appointments={extractAppointments(
                      appointmentResults[getStateOfAttendee(attendee).id]
                        .appointments
                    )}
                    selectedDate={selectedDate}
                    distanceFromLeft={(100 * index) / keyArray.length + 10}
                    name={getStateOfAttendee(attendee).name}
                    attendee={attendee}
                  />
                ))}
            </TimeTable>
            {visibilityService.getMeetingExists() ? (
              <PrStatusActionButton
                label={intl.formatMessage({
                  id: 'meetingcreator.termindetail'
                })}
                releaseButtonClick={handleChange}
              />
            ) : null}
          </Grid>
        ) : null}
      </Grid>
    </React.Fragment>
  );
};

export default withStyles(styles)(MeetingCreator);
