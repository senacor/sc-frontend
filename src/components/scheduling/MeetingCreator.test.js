import React from 'react';
import { MeetingCreator } from './MeetingCreator';
import { shallow } from 'enzyme';
import moment from 'moment';

describe('MeetingCreatorForm', () => {
  const pr = {
    id: 1,
    employee: {
      id: 1,
      firstName: 'Max',
      lastName: 'Mustermann',
      login: 'mmustermann',
      dateOfLastPr: '2019-03-14'
    },
    supervisor: {
      id: 3,
      firstName: 'Michaela',
      lastName: 'Mitarbeiterin',
      login: 'test.pr.mitarbeiter1'
    },
    dateOfLastPr: '2019-08-30',
    occasion: 'ON_DEMAND',
    statuses: ['FILLED_SHEET_EMPLOYEE', 'FILLED_SHEET_REVIEWER'],
    deadline: '2019-03-14',
    _links: {
      self: {
        href: 'http://localhost:8010/api/v1/prs/1'
      }
    }
  };

  const userinfo = { userPrincipalName: 'mmustermann' };

  const appointmentsSearchMock = jest.fn();
  const appointmentsSearchResults = {};

  beforeEach(() => {
    appointmentsSearchMock.mockClear();
  });

  it('should match snapshot', () => {
    let component = shallow(
      <MeetingCreator
        prDetail={pr}
        appointmentsSearchResults={appointmentsSearchResults}
        appointmentsSearch={appointmentsSearchMock}
        userinfo={userinfo}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should display the subcomponent MeetingCreatorForm ', () => {
    let component = shallow(
      <MeetingCreator
        prDetail={pr}
        appointmentsSearchResults={appointmentsSearchResults}
        appointmentsSearch={appointmentsSearchMock}
        userinfo={userinfo}
      />
    );

    expect(
      component.find('Connect(WithStyles(MeetingCreatorForm))').exists()
    ).toBeTruthy();
  });

  it('changes the visibility/PersonToggle of the attendee if onVisibilityChange is called.', () => {
    let component = shallow(
      <MeetingCreator
        prDetail={pr}
        appointmentsSearchResults={appointmentsSearchResults}
        appointmentsSearch={appointmentsSearchMock}
        userinfo={userinfo}
      />
    );

    component.instance().onVisibilityChange('supervisor')();

    expect(component.instance().state.supervisor.show).toBe(false);
  });

  it('fetches the appointments on mount', () => {
    let selectedDate = moment()
      .local()
      .format('YYYY-MM-DD');

    shallow(
      <MeetingCreator
        prDetail={pr}
        appointmentsSearchResults={appointmentsSearchResults}
        appointmentsSearch={appointmentsSearchMock}
        selectedDate={selectedDate}
        userinfo={userinfo}
      />
    );

    expect(appointmentsSearchMock).toHaveBeenCalledTimes(1);
    expect(appointmentsSearchMock.mock.calls[0][0]).toBe(
      `${pr.employee.login},${pr.supervisor.login}`
    );
    expect(appointmentsSearchMock.mock.calls[0][1]).toBe(
      moment()
        .local()
        .format('YYYY-MM-DD')
    );
  });

  it('takes the employee and supervisor from the PR and saves it to its state', () => {
    let component = shallow(
      <MeetingCreator
        prDetail={pr}
        appointmentsSearchResults={appointmentsSearchResults}
        appointmentsSearch={appointmentsSearchMock}
        userinfo={userinfo}
      />
    );
    const expectedComponentState = {
      employee: {
        id: pr.employee.login,
        name: `${pr.employee.firstName} ${pr.employee.lastName}`,
        role: 'Ich',
        show: true
      },
      supervisor: {
        id: pr.supervisor.login,
        name: `${pr.supervisor.firstName} ${pr.supervisor.lastName}`,
        role: 'Vorgesetzte/r',
        show: true
      }
    };

    expect(component.instance().state).toEqual(expectedComponentState);
  });
});
