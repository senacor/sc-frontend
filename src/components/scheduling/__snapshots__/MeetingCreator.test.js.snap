// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetingCreatorForm should match snapshot 1`] = `
<React.Fragment>
  <WithStyles(Typography)
    gutterBottom={true}
    variant="display1"
  >
    Terminfindung
  </WithStyles(Typography)>
  <WithStyles(Grid)
    container={true}
    direction="column"
    id="tableRolePick"
    spacing={24}
  >
    <WithStyles(Grid)
      item={true}
    >
      <Connect(WithStyles(MeetingCreatorForm))
        fetchAppointments={[Function]}
        prById={
          Object {
            "_links": Object {
              "self": Object {
                "href": "http://localhost:8010/api/v1/prs/1",
              },
            },
            "dateOfLastPr": "2019-08-30",
            "deadline": "2019-03-14",
            "employee": Object {
              "dateOfLastPr": "2019-03-14",
              "firstName": "Max",
              "id": 1,
              "lastName": "Mustermann",
              "login": "mmustermann",
            },
            "id": 1,
            "occasion": "ON_DEMAND",
            "statuses": Array [
              "FILLED_SHEET_EMPLOYEE",
              "FILLED_SHEET_REVIEWER",
            ],
            "supervisor": Object {
              "firstName": "Michaela",
              "id": 3,
              "lastName": "Mitarbeiterin",
              "login": "test.pr.mitarbeiter1",
            },
          }
        }
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
    >
      <WithStyles(Grid)
        container={true}
        direction="column"
      >
        <WithStyles(Grid)
          item={true}
          key="employee"
        >
          <PersonToggle
            displayName="mmustermann"
            displayRole="Ich"
            onChange={[Function]}
            showAttendee={true}
          />
        </WithStyles(Grid)>
        <WithStyles(Grid)
          item={true}
          key="supervisor"
        >
          <PersonToggle
            displayName="test.pr.mitarbeiter1"
            displayRole="Vorgesetzter"
            onChange={[Function]}
            showAttendee={true}
          />
        </WithStyles(Grid)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
    >
      <WithStyles(TimeTable) />
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</React.Fragment>
`;
